### PR TITLE
Homework SHM #1

### PR DESCRIPTION
**Pull request by arafalski**
Authors: @Vicx95, @kontrybutor, @piezol, @kari0036, @anraf1001

### CODE REVIEW

File: shm/inc/player.hpp [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    Player(Ship& ship, size_t money, size_t availableSpace)
+        : ship_(std::make_unique<Ship>(ship)), money_(money), availableSpace_(availableSpace) {}
+
+    std::unique_ptr<Ship> getShip() { return std::move(ship_); }
+    size_t getMoney() const { return money_; }
+    size_t getavailableSpace() const { return availableSpace_; }
+
+    size_t getSpeed() const { return ship_->getSpeed(); }
+
+    Cargo getCargo(size_t index) const { return ship_->getCargo(index); }
```
Comment (Pawellos) [inline shm/inc/player.hpp]: I think it supposed to be(return pointer) Cargo* getCargo(size_t index) const;

File: shm/inc/island.hpp [hunk 2]
```diff
@@ -0,0 +1,14 @@
+#pragma once
+
+#include "coordinates.hpp"
+
+class Island {
+public:
+    Island(int positionX, int positionY)
+        : position_(Coordinates(positionX, positionY)) {}
```
Comment (nauka-programowania-MA) [inline shm/inc/island.hpp]: move to .cc

File: shm/inc/cargo.hpp [hunk 3]
```diff
@@ -0,0 +1,23 @@
+#pragma once
+
+#include <string>
+
+class Cargo {
+public:
+    Cargo(const std::string& name, size_t amount, size_t basePrice)
+        : name_(name), amount_(amount), basePrice_(basePrice) {}
```
Comment (nauka-programowania-MA) [inline shm/inc/cargo.hpp]: move to .cc

File: shm/inc/island.hpp (hunk 2 above)
Comment (nauka-programowania-MA) [inline shm/inc/island.hpp]: move to cc

File: shm/inc/player.hpp [hunk 4]
```diff
@@ -0,0 +1,26 @@
+#pragma once
+
+#include <memory>
+
+#include "ship.hpp"
+
+class Player {
+public:
+    Player(Ship& ship, size_t money, size_t availableSpace)
+        : ship_(std::make_unique<Ship>(ship)), money_(money), availableSpace_(availableSpace) {}
```
Comment (nauka-programowania-MA) [inline shm/inc/player.hpp]: move to cc

File: shm/inc/player.hpp (hunk 1 above)
Comment (nauka-programowania-MA) [inline shm/inc/player.hpp]: Check index. And also return Cargo*

File: shm/inc/ship.hpp [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+
+class Ship {
+public:
+    Ship()
+        : id_(-1) {}
+
+    Ship(int capacity, int maxCrew, int speed, const std::string& name, unsigned int id)
+        : capacity_(capacity), maxCrew_(maxCrew), crew_(0), speed_(speed), name_(name), id_(id) {}
+    Ship(int maxCrew, int speed, unsigned int id)
+        : Ship(0, maxCrew, speed, "", id) {}
```
Comment (nauka-programowania-MA) [inline shm/inc/ship.hpp]: move to cc

File: shm/src/ship.cpp [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+Ship& Ship::operator-=(size_t num) {
+    crew_ -= num;
+    return *this;
+}
+
+Ship& Ship::operator+=(size_t num) {
+    crew_ += num;
+    return *this;
+}
```
Comment (nauka-programowania-MA) [inline shm/src/ship.cpp]: Check crew. TO not be too much and also dont drop below 0

File: shm/src/cargo.cpp [hunk 7]
```diff
@@ -0,0 +1,17 @@
+#include "cargo.hpp"
+
+Cargo& Cargo::operator+=(size_t amount) {
+    amount_ += amount;
+    return *this;
+}
+
+Cargo& Cargo::operator-=(size_t amount) {
+    amount_ -= amount;
+    return *this;
```
Comment (nauka-programowania-MA) [inline shm/src/cargo.cpp]: Check cargo. If not drop below 0

**Review by nauka-programowania-MA — CHANGES_REQUESTED** [review]

File: shm/src/cargo.cpp [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+    amount_ += amount;
+    return *this;
+}
+
+Cargo& Cargo::operator-=(size_t amount) {
+    if (amount_ > 0) {
+        amount_ -= amount;
+    }
+    return *this;
+}
```
Comment (Kasperice) [inline shm/src/cargo.cpp]: It will be worth to inform user what is going on if operation is not working

File: shm/src/ship.cpp [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    }
+    return *this;
+}
+
+Ship& Ship::operator+=(size_t num) {
+    if (crew_ < maxCrew_) {
+        crew_ += num;
+    }
+    return *this;
+}
```
Comment (Kasperice) [inline shm/src/ship.cpp]: You can inform user why operation was not done

File: shm/src/cargo.cpp [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    : name_(name), amount_(amount), basePrice_(basePrice) {}
+
+Cargo& Cargo::operator+=(size_t amount) {
+    amount_ += amount;
+    return *this;
+}
+
+Cargo& Cargo::operator-=(size_t amount) {
+    if (amount_ > 0) {
+        amount_ -= amount;
```
Comment (Marcinati) [inline shm/src/cargo.cpp]: If amount will be greater than amount_ it will go out or range of size_t.

File: shm/src/ship.cpp [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+
+Ship::Ship(int capacity, int maxCrew, int speed, const std::string& name, unsigned int id)
+    : capacity_(capacity), maxCrew_(maxCrew), crew_(0), speed_(speed), name_(name), id_(id) {}
+
+Ship::Ship(int maxCrew, int speed, unsigned int id)
+    : Ship(0, maxCrew, speed, "", id) {}
+
+Ship& Ship::operator-=(size_t num) {
+    if (crew_ > 0) {
+        crew_ -= num;
```
Comment (Marcinati) [inline shm/src/ship.cpp]: Same here.

File: shm/src/map.cpp [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+                               [&](const auto& island) {
+                                   return coordinates == island.getPosition();
+                               });
+    return island != std::end(islands_) ? &*island : nullptr;
+}
+
+void Map::generateIslands() {
+    std::random_device rd;
+    std::mt19937 eng(rd());
+    islands_.reserve(kIslandAmt);
```
Comment (Marcinati) [inline shm/src/map.cpp:27]: You can use: https://en.cppreference.com/w/cpp/numeric/random/uniform_int_distribution
As described in the task.

File: shm/src/ship.cpp [hunk 13]
```diff
@@ -0,0 +1,50 @@
+#include "ship.hpp"
+
+#include <iostream>
+#include <numeric>
```
Comment (WBigaj) [inline shm/src/ship.cpp:4]: Here you have included headers,then standard library includes.

File: shm/inc/ship.hpp [hunk 14]
```diff
@@ -0,0 +1,37 @@
+#pragma once
+
+#include <vector>
+
+#include "cargo.hpp"
```
Comment (WBigaj) [inline shm/inc/ship.hpp:5]: Here you have included standard library first,then headers.

**Review by WBigaj — COMMENTED** [review]
I have noticed that style of your includes is not coherent across the project. Generally we are trying to choose one formatting style and use it consistently.

File: shm/src/ship.cpp (hunk 13 above)
Comment (arafalski) [reply shm/src/ship.cpp:4]: Firstly we include header file that has declaration of class. Then standard libraries. Last but not least - additional headers that need to be used in that file.

Comment (arafalski) [reply shm/src/ship.cpp:4]: So I think that it is consistent

**Review by nauka-programowania-MA — APPROVED** [review]

**Comment by nauka-programowania-MA** [comment]
{
"requirement_ids": [2568, 2568, 2568, 2568, 2568, 2568, 2568, 2568, 2569, 2570],
"points": [8, 8, 8, 8, 8, 8, 8, 8, 20, 6],
"users": ["Vicx95", "kontrybutor", "kari0036", "piezol", "anraf1001"]
}

**Pull request closed without merging**